man/pione-clean.1
.TH pione\-clean 1 "" "PIONE User Manual"
.SH NAME
.PP
pione\-clean \- remove PIONE\[aq]s temporary files, cache, and etc.
.SH SYNOPSIS
.PP
\f[B]pione\-clean\f[] [options]
.SH DESCRIPTION
.PP
pione\-clean removes PIONE\[aq]s temporary files, cache, and etc.
.SS NOTE
.PP
Package cache files are not removed if it is in package database.
.SH OPTIONS
.TP
.B \-\-older=\f[I]DATE\f[]
Remove files older than the date.
\f[I]DATE\f[] can be iso8601 format:
.RS
.IP
.nf
\f[C]
pione\-clean\ \-\-older=2013\-10\-01
\f[]
.fi
.PP
This means pione\-clean command removes files older than 2013\-10\-01.
Otherwise it can be days:
.IP
.nf
\f[C]
pione\-clean\ \-\-older=10
\f[]
.fi
.PP
This means pione\-clean command removes files older than 10 days.
.RE
.TP
.B \-\-type=\f[I]NAME\f[]
Remove only files of the type.
\f[I]NAME\f[] can be \f[I]temporary\f[], \f[I]file\-cache\f[],
\f[I]package\-cache\f[], or \f[I]profile\f[].
.RS
.RE
.SH EXAMPLES
.TP
.B pione\-clean
Remove all type files.
.RS
.RE
.TP
.B pione\-clean \-\-older=2013\-10\-01
Remove all type files older than 2013\-10\-01.
.RS
.RE
.TP
.B pione\-clean \-\-older=10
Remove all type files older than 10 days.
.RS
.RE
.TP
.B pione\-clean \-\-type=profile
Remove profile reports.
.RS
.RE
.TP
.B pione\-clean \-\-type=file\-cache \-\-older=30
Remove file cache files older than 30 days.
.RS
.RE
.SH REPORTING BUGS
.PP
Report bugs or feature requests to PIONE\[aq]s issue
tracker(https://github.com/pione/pione/issues).
.SH AUTHORS
Keita Yamaguchi.


History.txt
# History

## 0.3.1(2013/10/29)

* Fixed performance bug.
* Enable to generate profile reports of rule application.
* Added PNML compiler.
* Changed to stop processing when action rule happens error.
* Improved pione-clean command and the document, see `man pione-clean`.
* Added `pione` facade command.

## 0.3.0(2013/10/20)

* Improved multi-inheritance package system.
* Added package database.
* Added package cache system.
* Improved language parser and interpreter.
* Refactored many codes.
* Improved test coverage.

## 0.2.2(2013/07/02)

* Enabled to load remote package.
* Added pione-package command.
* Added PIONE archive(".ppg" file).
* Added git package.
* Fixed the problem about pione-client termination.
* Refactored location system.
* Added HTTPS location.
* Added examples.

## 0.2.1(2013/06/15)

* Added basic/advaneced parameter.
* Enabled to run rehearsal test.
* Imporved test coverage.

## 0.2.0(2013/05/31)

* Added initial package system.
* Added pione-val.
* Improved sequence.
* Improved message log framework.
* Added system log framework.
* Changed to unify process log.
* Added endurance test scripts.
* Added embedded FTP server.
* Reduced OS dependencies.
* Added examples of WorkflowPatterns.

## 0.1.4(2013/05/13)

* Sequence based expressions.
* Enable to distribute tasks based on parameters.
* Added many methods.
* Added annotations.
* Added remove and touch operation as data tuple operations.
* Modified update criteria.
* Added empty rule.
* Added new examples.

## 0.1.3(2013/04/25)

* Added new data expression literal "null" for accepting data nonexistence
* Added new data expression property keyword "neglect" and "care" for the update criterion about data mtime
* Changed log location to be in output location
* Changed log format to JSON
* Added pione-log command that generates agent activity log as XES format
* Use locations in tuples
* Enabled to test FTP location handling without setting up FTP server
* Added the feature to create domain information files(".domain_info")

## 0.1.2(2013/04/01)

* Added shorthand notation for setting parameters
* Added ticket concept for sequential processing
* Improved termination sequence
* Refactored parser and transformer

## 0.1.1(2013/03/20)

* Improved performance
* Added the concept of document parameter
* Added --list-params option
* Fixed bugs of features handling
* Improved comment parser

## 0.1.0(2013/02/19)

* First gem release


lib/pione/command/pione-clean.rb
module Pione
  module Command
    # PioneClean is a command for clearing temporary files created by PIONE
    # system.
    class PioneClean < BasicCommand
      #
      # basic informations
      #

      command_name "pione-clean"
      command_banner "remove PIONE's temporary files, cache, and etc."
      PioneCommand.add_subcommand("clean", self)

      #
      # options
      #

      use_option :debug

      # --older option
      define_option(:older) do |item|
        item.long = '--older=DATE'
        item.desc = 'remove file older than the date'
        item.default = false
        item.value = proc do |str|
          begin
            Date.iso8601(str)
          rescue ArgumentError
            begin
              Date.today - str.to_i
            rescue
              abort("invalid value of older option: %s" % str)
            end
          end
        end
      end

      # --type option
      define_option(:type) do |item|
        item.long = '--type=NAME'
        item.desc = 'remove only files of the type'
        item.default = false
        item.value = proc do |name|
          case name
          when "temporary"
            :temporary
          when "file-cache"
            :file_cache
          when "package-cache"
            :package_cache
          when "profile"
            :profile
          else
            abort("unknown type: %s" % str)
          end
        end
      end

      #
      # command lifecycle: setup phase
      #

      setup :package_database

      def setup_package_database
        @db = Package::Database.load
      end

      #
      # command lifecycle: execution phase
      #

      execute :remove_temporary
      execute :remove_file_cache
      execute :remove_package_cache
      execute :remove_profile

      # Remove temporary files. This removes working files based on temporary
      # directory's mtime.
      def execute_remove_temporary
        if type?(:temprary)
          Location[Global.my_temporary_directory].each_entry do |entry|
            if delete?(entry)
              FileUtils.remove_entry_secure(entry.path)
            end
          end
        end
      end

      # Remove file cache files. This removes it base on cache directory's mtime.
      def execute_remove_file_cache
        if type?(:file_cache)
          Location[Global.my_file_cache_directory].each_entry do |entry|
            if delete?(entry)
              FileUtils.remove_entry_secure(entry.path)
            end
          end
        end
      end

      # Remove package cache files.
      def execute_remove_package_cache
        if type?(:package_cache)
          # remove PPG package
          Global.ppg_package_cache_directory.each_entry do |entry|
            unless @db.has_digest?(Package::PackageFilename.parse(entry.basename).digest)
              if delete?(entry)
                entry.delete
              end
            end
          end

          # remove directory package
          Global.directory_package_cache_directory.each_entry do |entry|
            unless @db.has_digest?(entry.basename)
              if delete?(entry)
                entry.delete
              end
            end
          end
        end
      end

      # Remove profile reports.
      def execute_remove_profile
        if type?(:profile)
          Location[Global.profile_report_directory].each_entry do |entry|
            if delete?(entry)
              entry.delete
            end
          end
        end
      end

      #
      # helper methods
      #

      # Return true if the entry has the type.
      def type?(type)
        option[:type].nil? or option[:type] == type
      end

      # Return true if the entry should be removed.
      def delete?(entry)
        option[:older].nil? or option[:older] >= entry.mtime.to_date
      end
    end
  end
end
